Add snippets to GHA (#1069)
localstack/docs: content/en/user-guide/ci/github-actions/samples/index.md

---
title: "Sample"
linkTitle: "Sample"
weight: 5
description: >
  Localstack Sample Project in Github Actions
aliases:
  - /ci/github-actions/samples
tags: ["continuous-integration", "ci", "continuous-delivery", "testing","github-actions","sample"]
---

This page containes our Quickstart project's test pipeline implemented in Github actions.

# Serverless image resizer s3 lambda
This is an app to resize images uploaded to S3 in a serverless way.
Find code in repo [here](https://github.com/localstack-samples/sample-serverless-image-resizer-s3-lambda).

## Integration testing
A build matrix for Ubuntu and MacOS to run integration tests triggered by various ways.

> In case of failure the pipeline dumps the logs of Localstack and preserves them as artifacts for further investigations.

Reviewer: That's actually not true, the worfklow currently has a `if: always()` for the `Show localstack logs`. 
If you were refering to the diagnostic file, we should rephrase this part :) Also in that case it would be worth mentioning that the flag `LS_LOG=trace` needs to be set for the `setup-localstack` action.

```
name: Run Integration Tests

on:
  push:
    paths-ignore:
      - 'README.md'
    branches:
      - main
  pull_request:
    branches:
      - main
  schedule:
    # “At 00:00 on Sunday.”
    - cron: "0 0 * * 0"

>   workflow_dispatch:

Reviewer: not sure if it makes sense to keep all of this in the docs. it adds additional noise with no real value. 
If we keep it, we should at least explain what it does (adding inline comments would be sufficient).

        uses: LocalStack/setup-localstack@main
        with:
          image-tag: 'latest'
          use-pro: 'true'
          configuration: LS_LOG=trace
          install-awslocal: 'true'
        env:
          LOCALSTACK_API_KEY: ${{ secrets.LOCALSTACK_API_KEY }}

      - name: Deploy infrastructure
        run: |
          bin/deploy.sh

      - name: Run Tests
        env:
          AWS_DEFAULT_REGION: us-east-1
          AWS_REGION: us-east-1
          AWS_ACCESS_KEY_ID: test
          AWS_SECRET_ACCESS_KEY: test
        run: |
          pytest tests

      - name: Show localstack logs
        if: always()
        run: |
          localstack logs

      - name: Send a Slack notification
        if: failure() || github.event_name != 'pull_request'
        uses: ravsamhq/notify-slack-action@v2
        with:
          status: ${{ job.status }}
          token: ${{ secrets.GITHUB_TOKEN }}
          notification_title: "{workflow} has {status_message}"
          message_format: "{emoji} *{workflow}* {status_message} in <{repo_url}|{repo}>"
          footer: "Linked Repo <{repo_url}|{repo}> | <{run_url}|View Workflow run>"
          notify_when: "failure"
        env:

>           SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}

Reviewer: I think we should remove this from the docs.

      - name: Generate a Diagnostic Report
        if: failure()
        run: |
          curl -s localhost:4566/_localstack/diagnose | gzip -cf > diagnose.json.gz

      - name: Upload the Diagnostic Report
        if: failure()
        uses: actions/upload-artifact@v3
        with:
          name: diagnose.json.gz
          path: ./diagnose.json.gz
```

## Cloud pod testing

> Cloud pods providing a quick and easy way to test newly written tests against existing infrastructure or deployed code.

Reviewer: ```suggestion
Cloud pods provide a quick and easy way to test newly written tests against existing infrastructure or deployed code.
```

> This can be stored locally, like we did below, or remotely in our Web Application.

Reviewer: ```suggestion
A cloud pod be stored locally, like we did below, or remotely in our Web Application.
```

> Find more information on them [here](/user-guide/state-management/cloud-pods/).

Reviewer: ```suggestion
You can find detailed information about cloudpods in our [guide here](/user-guide/state-management/cloud-pods/).
```

> ### Create pods

Reviewer: I don't think this is a good example TBH. 
IMO it would make more sense to create a smaller sample that only creates a pod, without all the release specific details. 
It would IMO be better to create a showcase workflow that creates a pod and uploads it as an artifact.

> Preserves infrastructure and deployed code for different versions.

Reviewer: i have troubles understanding this sentence :) 
I would suggest something like:

> You can create a pod after setting up your specific infrastructure. The pod can later on be injected in a LocalStack instance, restoring the original infrastructure. In this sample we deploy the sample app and run the tests, before creating the pod. 

If we change the workflow I would also add:

>  The pod will be stored as an artifact and can be downloaded.

        run: |
          echo "This release includes the Cloud Pod of the sample created with LocalStack Version \`${{ inputs.release-tag || 'latest'}}\`." > Release.txt
          echo "You can download the \`release-pod.zip\` and inject it manually by running \`localstack state import release-pod.zip\`, or use the Cloud Pods Launchpad." >> Release.txt
          echo "### Cloud Pods Launchpad" >> Release.txt
          echo "You can click the Launchpad to inject the the pod into your running LocalStack instance using the WebUI:" >> Release.txt
          echo "[![LocalStack Pods Launchpad](https://localstack.cloud/gh/launch-pod-badge.svg)](https://app.localstack.cloud/launchpad?url=https://github.com/$GITHUB_REPOSITORY/releases/download/${{ inputs.release-tag || 'latest'}}/release-pod.zip)" >> Release.txt

      - name: Create Release
        id: create_release
        uses: softprops/action-gh-release@v1
        with:
          tag_name: "${{ inputs.release-tag || 'latest'}}"
          name: "Cloud Pod for LocalStack Version '${{ inputs.release-tag || 'latest'}}'"
          body_path: ./Release.txt
          files: |
              ./release-pod.zip
```

> ### Test pods

Reviewer: Same as above: I don't think this a good example. 
I would remove all the release-specific parts and create a new sample workflow. We could run additional, or different tests there to make the sample more concise and argument why this can be useful for developers.

The workflow that we have in the repo mainly helps us (LocalStack) to get notified if a new LS version breaks an existing pod.

        if: failure()
        run: |
          localstack logs

      - name: Send a Slack notification
        if: failure() || github.event_name != 'pull_request'
        uses: ravsamhq/notify-slack-action@v2
        with:
          status: ${{ job.status }}
          token: ${{ secrets.GITHUB_TOKEN }}
          notification_title: "{workflow} has {status_message}"
          message_format: "{emoji} *{workflow}* {status_message} in <{repo_url}|{repo}>"
          footer: "Linked Repo <{repo_url}|{repo}> | <{run_url}|View Workflow run>"
          notify_when: "failure"
        env:
          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}

      - name: Prevent Workflows from getting Stale
        if: always()
        uses: gautamkrishnar/keepalive-workflow@v1
        with:
          # this message should prevent automatic triggering of workflows
          # see https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs
          commit_message: "[skip ci] Automated commit by Keepalive Workflow to keep the repository active"
```

## Preview application
Application Preview allows you to generate an preview environment from GitHub Pull Request (PR) builds.

> To communicate with the ephemeral instance use it's address as an AWS endpoint URL from your local environment.

Reviewer: ```suggestion
To communicate with the ephemeral instance use its address as an AWS endpoint URL from your local environment.
```

For more information please read the related section [here](https://docs.localstack.cloud/user-guide/cloud-sandbox/application-previews/).

> ### Create preview

Reviewer: Nice 👍

The pipeline creates a preview environment by deploying the application for every pull requests.
```
name: Create PR Preview

on:
  pull_request:
    types: [opened, synchronize, reopened]

jobs:
  test:
    runs-on: ubuntu-latest
    timeout-minutes: 15
    permissions:
      pull-requests: write
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Set up Python 3.9
        id: setup-python
        uses: actions/setup-python@v2
        with:
          python-version: 3.9

      - name: Install dependencies
        run: |
          pip install awscli-local

      - name: Deploy Preview
        uses: LocalStack/setup-localstack/preview@main
        env:
          AWS_DEFAULT_REGION: us-east-1
          AWS_REGION: us-east-1
          AWS_ACCESS_KEY_ID: test
          AWS_SECRET_ACCESS_KEY: test
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          localstack-api-key: ${{ secrets.LOCALSTACK_API_KEY }}
          preview-cmd: |
            # Add your custom deployment commands here. 
            # Below is an example for the Image resizer application.
            bin/deploy.sh  
```

### Attach preview URL

> It attaches the Preview URL to the pull request.

Reviewer: ```suggestion
This workflow showcases how to attach the Preview URL to the pull request.
```
